src/fs/fuse.rs

```diff
@@ -38,1190 +38,557 @@ const U64_SIZE: usize = ::core::mem::size_of::<u64>();
-struct fuse_out_header {
-	pub len: u32,
-	pub error: i32,
-	pub unique: u64,
-}
+/// Marker trait, which signals that a struct is a valid Fuse command.
+/// Struct has to be repr(C)!
+pub(crate) unsafe trait FuseIn {}
+/// Marker trait, which signals that a struct is a valid Fuse response.
+/// Struct has to be repr(C)!
+pub(crate) unsafe trait FuseOut {}
 
-#[repr(C)]
-#[derive(Debug, Default)]
-struct fuse_init_in {
-	pub major: u32,
-	pub minor: u32,
-	pub max_readahead: u32,
-	pub flags: u32,
+unsafe impl FuseIn for fuse_abi::InitIn {}
+unsafe impl FuseOut for fuse_abi::InitOut {}
+unsafe impl FuseIn for fuse_abi::ReadIn {}
+unsafe impl FuseIn for fuse_abi::WriteIn {}
+unsafe impl FuseOut for fuse_abi::WriteOut {}
+unsafe impl FuseOut for fuse_abi::ReadOut {}
+unsafe impl FuseIn for fuse_abi::LookupIn {}
+unsafe impl FuseIn for fuse_abi::ReadlinkIn {}
+unsafe impl FuseOut for fuse_abi::ReadlinkOut {}
+unsafe impl FuseOut for fuse_abi::AttrOut {}
+unsafe impl FuseOut for fuse_abi::EntryOut {}
+unsafe impl FuseIn for fuse_abi::CreateIn {}
+unsafe impl FuseOut for fuse_abi::CreateOut {}
+unsafe impl FuseIn for fuse_abi::OpenIn {}
+unsafe impl FuseOut for fuse_abi::OpenOut {}
+unsafe impl FuseIn for fuse_abi::ReleaseIn {}
+unsafe impl FuseOut for fuse_abi::ReleaseOut {}
+unsafe impl FuseIn for fuse_abi::RmdirIn {}
+unsafe impl FuseOut for fuse_abi::RmdirOut {}
+unsafe impl FuseIn for fuse_abi::MkdirIn {}
+unsafe impl FuseIn for fuse_abi::UnlinkIn {}
+unsafe impl FuseOut for fuse_abi::UnlinkOut {}
+unsafe impl FuseIn for fuse_abi::LseekIn {}
+unsafe impl FuseOut for fuse_abi::LseekOut {}
+unsafe impl FuseIn for fuse_abi::PollIn {}
+unsafe impl FuseOut for fuse_abi::PollOut {}
+
+pub(crate) trait OpTrait {
+	type InStruct: FuseIn + core::fmt::Debug;
+	type InPayload: ?Sized;
+	type OutStruct: FuseOut + core::fmt::Debug;
+	type OutPayload: ?Sized;
+
+	type Cmd: ?Sized + AsSliceU8 = ReqPart<fuse_abi::InHeader, Self::InStruct, Self::InPayload>;
+	type Rsp: ?Sized + AsSliceU8 =
+		ReqPart<fuse_abi::OutHeader, MaybeUninit<Self::OutStruct>, Self::OutPayload>;
 }
 
-unsafe impl FuseIn for fuse_init_in {}
+pub(crate) struct Op<const CODE: u32>;
 
-#[repr(C)]
-#[derive(Debug, Default)]
-struct fuse_init_out {
-	pub major: u32,
-	pub minor: u32,
-	pub max_readahead: u32,
-	pub flags: u32,
-	pub max_background: u16,
-	pub congestion_threshold: u16,
-	pub max_write: u32,
-	pub time_gran: u32,
-	pub unused: [u32; 9],
+impl OpTrait for Op<{ fuse_abi::Opcode::Init as u32 }> {
+	type InStruct = fuse_abi::InitIn;
+	type InPayload = ();
+	type OutStruct = fuse_abi::InitOut;
+	type OutPayload = ();
 }
-unsafe impl FuseOut for fuse_init_out {}
 
-#[repr(C)]
-#[derive(Default, Debug)]
-struct fuse_read_in {
-	pub fh: u64,
-	pub offset: u64,
-	pub size: u32,
-	pub read_flags: u32,
-	pub lock_owner: u64,
-	pub flags: u32,
-	pub padding: u32,
+impl Op<{ fuse_abi::Opcode::Init as u32 }> {
+	fn create() -> (Box<<Self as OpTrait>::Cmd>, Box<<Self as OpTrait>::Rsp>) {
```

Comment:
For `Sized` return types, we might want to avoid boxing if possible. Though that might be better off as a separate PR, since boxing was the previous behavior too, I think.